### DISCUSSION TITLE
OPENSSL_init_crypto usage in OpenSSL 3.0

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#22686] [Q&A]
when do we use OPENSSL_init_crypto ? Is it must to use this for using openssl functions ? I currently don't use it and all functionality works fine for me . In what scenarios is it must to use OPENSSL_init_crypto

**Comment by mattcaswell** [comment]
You typically do not need to call it at all.

From the man page:

> As of version 1.1.0 OpenSSL will automatically allocate all resources that it needs so no explicit initialisation is required. Similarly it will also automatically deinitialise as required.
> 
> However, there may be situations when explicit initialisation is desirable or needed, for example when some nondefault initialisation is required. The function OPENSSL_init_crypto() can be used for this purpose for libcrypto (see also [OPENSSL_init_ssl(3)](https://www.openssl.org/docs/manmaster/man3/OPENSSL_init_ssl.html) for the libssl equivalent).

https://www.openssl.org/docs/manmaster/man3/OPENSSL_init_crypto.html

There are some options to `OPENSSL_initi_crypto()` which are not the default way of initialising OpenSSL. For example the `OPENSSL_INIT_NO_LOAD_CONFIG` option suppresses loading the config file. If you want to use one of these options then you must call `OPENSSL_init_crypto()` explicitly.

Some applications call `OPENSSL_init_crypto()` to force initialisation to happen "right now" - as opposed to lazily on an as needed basis. For that to work you have to pass all the options for the things that you want to be initialised now.

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

Thanks for your response.

I am using openssl non-shared libs into my project but using these it was pinning dll having the openssl libs and thus interfering with our dll unloading logic. 

What i want to know is . Is it safe to use [no-pinshared](https://github.com/openssl/openssl/blob/master/INSTALL.md#no-pinshared) for non-shared libraries ? And as per my understanding this pinning logic of dll was specifically for shared libs only . please confirm

**Reply by mattcaswell** [reply]
If you are going to use no-pinshared either with shared or non-shared libraries you need to be careful about how OpenSSL handles exit. OpenSSL automatically registers an "atexit" handler. That handler is expected to still be in memory at the point that the process exits. If that is not the case then a crash is likely.

If it is possible that OpenSSL will no longer be in memory at process exit then you should use the `OPENSSL_INIT_NO_ATEXIT` initialisation option to `OPENSSL_init_crypto()`. Since this is a non-default option this is one of the cases where you need to call it explicitly.

The effect of that option is that OpenSSL will no longer register an atexit handler and will not clean itself up on exit. You must either explicitly clean it up by calling `OPENSSL_cleanup()` - or just live without any explicit cleanup. It does not harm anything to have no cleanup. The cleanup just releases various resources that OpenSSL was holding. All of those resources will be reclaimed anyway by the OS when the process exits so there is no leak. However, if you use any memory error detection software you may see false positive reports of leaks.

**Reply by aniljadaun** [reply]
Thanks @mattcaswell

**Reply by t8m** [reply]
Hm.... perhaps we should imply OPENSSL_INIT_NO_ATEXIT when no-pinshared build option is used?

**Reply by aniljadaun** [reply]
@mattcaswell @t8m We use fips workflow as well. This is work fine for non-fips workflow but now i am getting crash from fips.dll at exit .Do i need to explicitely unload fips.dll as well at the OPENSSL_cleanup ?

**Reply by t8m** [reply]
Do you call OPENSSL_init_crypto() with OPENSSL_INIT_NO_ATEXIT? Or does possibly the fips module register its own atexit handler @mattcaswell ?

**Reply by mattcaswell** [reply]
> Or does possibly the fips module register its own atexit handler @mattcaswell ?

No. It is cleaned up by the main atexit in libcrypto.

**Reply by mattcaswell** [reply]
> Do you call OPENSSL_init_crypto() with OPENSSL_INIT_NO_ATEXIT?

Right - you probably need to do this if you are using no-pinshared. It must be *the first thing you do with OpenSSL*

**Reply by aniljadaun** [reply]
@mattcaswell @t8m  What is the use of openssl_conf = openssl_init in the configuration file ? Does this also impact OPENSSL_INIT_NO_ATEXIT ?

What is recommended order for calling  OPENSSL_init_crypto with OPENSSL_INIT_NO_ATEXIT and loading fips.dll and setting it as 
default property with EVP_default_properties_enable_fips(NULL,1);

**Reply by aniljadaun** [reply]
Hi @mattcaswell , @t8m please suggest . i am stuck at this .

First thing i do with openssl is call OPENSSL_init_crypto with OPENSSL_INIT_NO_ATEXIT after that i load fips provider using conf file set fips as default property using EVP_default_properties_enable_fips at exit call EVP_default_properties_enable_fips with 0 to unset fips and do OPENSSL_cleanup in the end. 
Still i get crash in the end indicating crash from fips.dll!CRYPTO_free not sure why is that if along with OPENSSL_cleanup i explicitely unload fips.dll as well the crash is not seen. 

please suggest . i guess from your comment explicit unloading should not be  required. Can you please check if this is bug from openssl side.

I built libcrypto from 3.0.10 with inbuilt legacy and no-pinshared option, i built fips module from 3.0.8  with no-pinshared .

Content of openssl.cnf
---------------------------------------------
config_diagnostics = 1
openssl_conf = openssl_init
.include "path to fipsmodule.cnf"
[openssl_init]
providers = provider_sect

[provider_sect]
fips = fips_sect

[fips_sect]
identity=fips-customname

content of fipsmodule.cnf
---------------------------------------------
[fips_sect]
activate = 1
install-version = 1
conditional-errors = 1
security-checks = 1
module-mac = some_content
install-mac = some_content
install-status = INSTALL_SELF_TEST_KATS_RUN
